### PR TITLE
HLA-1191 GitHub Contributors Guide

### LINKED ISSUE #1732: New readthedocs Github contribution guide
_Issue [HLA-1191](https://jira.stsci.edu/browse/HLA-1191) was created on JIRA by [Steve Goldman](https://jira.stsci.edu/secure/ViewProfile.jspa?name=sgoldman):_

The contributor instructions on readthedocs are out-of-date. The dependencies page is also the developer dependencies for testing and not the full dependency list shown in the pyproject toml file.

 

We should create a new Github contributors page modeled after that of the [JWST pipeline](https://github.com/spacetelescope/jwst/blob/master/CONTRIBUTING.md).

### PR DESCRIPTION
**Pull request by s-goldman**
Resolves [HLA-1191](https://jira.stsci.edu/browse/HLA-1191)

Closes #1732 

This PR adds a new github contributors guide for basic and advanced experience levels. The document is based on the [contributors guide for JWST](https://github.com/spacetelescope/jwst/blob/master/CONTRIBUTING.md). I have also copied over the code of conduct from the jwst pipieline repository. 

**Checklist for maintainers**
- [x] added entry in `CHANGELOG.rst` within the relevant release section
- [x] updated or added relevant tests
- [x] updated relevant documentation
- [ ] added relevant milestone
- [x] added relevant label(s)
- [ ] ran regression tests, post a link to the Jenkins job below.
      [How to run regression tests on a PR](https://github.com/spacetelescope/jwst/wiki/Running-Regression-Tests-Against-PR-Branches)

### CODE REVIEW

File: CONTRIBUTING.rst [hunk 1]
```diff
@@ -0,0 +1,480 @@
...
+::
+
+   >> git remote add upstream git@github.com:spacetelescope/drizzlepac.git
+
+Otherwise, you can simply set it to the repository URL but you will have
+to enter your password every time you fetch/push
+
+::
+
+   >> git remote add upstream https://github.com/spacetelescope/drizzlepac
```
Comment (mdlpstsci) [inline CONTRIBUTING.rst:110]: I am not sure this change is actually needed, but for completeness the URL would be:
https://github.com/spacetelescope/drizzlepac.git

Comment (s-goldman) [reply CONTRIBUTING.rst]: fixed.

File: CONTRIBUTING.rst [hunk 2]
```diff
@@ -0,0 +1,480 @@
...
+
+Unit tests are located in the ``tests`` directory and are separated by instrument,
+with additional separate directories for Hubble Advanced Products (HAP) and the 
+drizzle algorithm. These tests run the code
+on simplified datasets to make sure there are no breaking changes
+introduced. We aim to cover most lines of ``drizzlepac`` with unit
+test, so when adding code you will often need to write a new test or add
+to an existing test to ensure adequate coverage.
+
+Take a look around at the existing tests for a template - a majority of
```
Comment (mdlpstsci) [inline CONTRIBUTING.rst:348]: I would re-word 348-354 as this is not really true of the drizzlepac tests.  A "datamodel" is not necessarily used and the simplified data/dq is also not usually true.  The HAP tests are real "as is" datasets.

File: CONTRIBUTING.rst [hunk 3]
```diff
@@ -0,0 +1,480 @@
...
+
+To run all of the ``drizzlepac`` unit tests, while in the
+``drizzlepac/`` level directory, simply run the command:
+
+::
+
+   >> pytest
+
+If you want to run all the tests for a single instrument/directory, for example
+``WFC3``, you can run this from 'drizzlepac/wfc3'.
```
Comment (mdlpstsci) [inline CONTRIBUTING.rst:388]: Fix: drizzlepac/tests/wfc3

File: CONTRIBUTING.rst [hunk 4]
```diff
@@ -0,0 +1,480 @@
...
+
+To run all tests within a single test file (for example, all tests in
+``test_wfc3``). 
+
+::
+   
+   >> pytest drizzlepac/wfc3/test_wfc3.py
```
Comment (mdlpstsci) [inline CONTRIBUTING.rst:395]: Fix: drizzlepac/tests/wfc3/test_wfc3.py

File: CONTRIBUTING.rst [hunk 5]
```diff
@@ -0,0 +1,480 @@
...
+run by the CI) 3. @pytest.fixture to declare a
+`fixture <https://docs.pytest.org/en/7.1.x/explanation/fixtures.html>`__
+1. @pytest.mark.xfail will make a test pass only if it fails.
+
+Simultaneously developing ``drizzlepac`` and one of its dependencies
+--------------------------------------------------------------------
+
+If you encounter the scenario where you wish to simultaneously make
+changes in ``drizzlepac`` and also in one of its dependencies like
+``stcal`` or ``stpipe``, we recommend that you create a new environment
```
Comment (mdlpstsci) [inline CONTRIBUTING.rst:437]: 'stcal' and 'stpipe' are not dependencies of 'drizzlepac'.  Pick something for illustration from pyproject.toml.

File: CONTRIBUTING.rst [hunk 6]
```diff
@@ -0,0 +1,480 @@
...
+its dependency packages, unit tests will not pass on the CI because the
+``pyproject.toml`` file in ``drizzlepac`` points to the last released
+version of ``stcal``, and stcal points to the last version of
+``drizzlepac``, so the issue becomes circular. What you will need to do
+is modify the ``pyproject.toml`` files in both packages to point to the
+other to demonstrate that CI tests pass (and make a comment noting this
+in your PR), and then change it back before the PR is merge so that
+changes to ``pyproject.toml`` are not merged into main. In your
+``drizzlepac`` branch, to point to your branch in the dependent package
+(in this example ``stcal``), change the required ``stcal`` version in
```
Comment (mdlpstsci) [inline CONTRIBUTING.rst:463]: Change references to 'stcal' to a package which is really a dependency of 'drizzlepac'.

**Review by mdlpstsci — CHANGES_REQUESTED** [review]
Please see comments which I think I only made in CONTRIBUTING.rst.  The comments are to update the boilerplate to items which apply to 'drizzlepac'.

**Comment by s-goldman** [comment]
I also fixed the requirements page that now points to a selection of the pyproject.toml file instead of the requirements.txt file.

File: CONTRIBUTING.rst [hunk 7]
```diff
@@ -0,0 +1,477 @@
...
+This will install the optional 'test' dependencies specified in
+``pyproject.toml`` that don't install by default. The package ``pytest``
+is one of these and is what's used to run the tests. ``pytest`` searches
+through all the directories in your repository (underneath the directory
+from which it was invoked command line) and looks for any directories
+called 'test' or .py files with the word 'test' in the name. Functions
+in these files will be executed.
+
+To run all of the ``drizzlepac`` unit tests, while in the
+``drizzlepac/`` level directory, simply run the command:
```
Comment (mdlpstsci) [inline CONTRIBUTING.rst:378]: This will only run the tests which are not marked as "slow".

File: CONTRIBUTING.rst [hunk 8]
```diff
@@ -0,0 +1,477 @@
...
+   >> pip install -e ".[test]"
+
+This will install the optional 'test' dependencies specified in
+``pyproject.toml`` that don't install by default. The package ``pytest``
+is one of these and is what's used to run the tests. ``pytest`` searches
+through all the directories in your repository (underneath the directory
+from which it was invoked command line) and looks for any directories
+called 'test' or .py files with the word 'test' in the name. Functions
+in these files will be executed.
+
```
Comment (mdlpstsci) [inline CONTRIBUTING.rst:376]: Steve, Put the following text where you think it is the most appropriate and reword as you want. Further down in this write-up may be better for this discussion where you talk about flag options.  You will see hints of this in the JenkinsfileRT.

The tests which import hst_helpers from the ci_watson package (e.g., hap, acs, wfc3) need to have the environment variable TEST_BIGDATA defined for the appropriate location in Artifactory (e.g., TEST_BIGDATA=https://bytesalad.stsci.edu/artifactory/) or directory path on your local disk where the _input_ and _truth_ files area located.  You also need to use the flag, --bigdata, to have the FITS files downloaded into the testing area.

Some tests are also marked as "slow" and will only be executed when the "--slow" flag is included on the "pytest" command line.  These tests are typically NOT executed as part of the automatically invoked regression test suite due the amount of time needed to execute the test.

**Review by mdlpstsci — CHANGES_REQUESTED** [review]
Since you are adding this text now, I have added a few lines to discuss the proper way to execute some of the tests (or they will not run).  We should discuss this at a later date so you can improve the text here.

File: CONTRIBUTING.rst (hunk 8 above)
Comment (s-goldman) [reply CONTRIBUTING.rst:376]: I was under the impression that this document was more for a possible non-STScI contributor, whereas our internal notes are on the innerspace page. Can a non-STScI user run the --slow and --bigdata tests? My impression was that they couldn't, but I may be wrong.

**Review by mdlpstsci — DISMISSED** [review]
With regard to my comments on "--slow" and "--bigdata"... ==> ignore my comments.

Outside users just need to know there will be some tests which will fail for them because they do not have access internal data.   Please just add a comment for outside users so they have a perspective about this.

**Review by mdlpstsci — APPROVED** [review]

**Pull request merged**